src/fd.cpp

```diff
@@ -0,0 +1,213 @@
+
+  // store pollfd revents in args->results for use by callback
+
+  if (ready > 0) {
+    for (int i = 0; i < args->num_fds; i++) {
+      (args->results)[i] = (args->fds)[i].revents == 0 ? 0 : (args->fds)[i].revents & (POLLIN | POLLOUT) ? 1: NA_INTEGER;
+    }
+  } else if (ready < 0) {
+    std::fill(args->results.begin(), args->results.end(), NA_INTEGER);
+  }
+
+  callbackRegistryTable.scheduleCallback(later_callback, static_cast<void *>(argsptr.release()), 0, args->loop);
+
+  return 0;
+
+}
+
+static int execLater_launch_thread(std::shared_ptr<ThreadArgs> args) {
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(new std::shared_ptr<ThreadArgs>(args));
+
+  tct_thrd_t thr;
+
+  return tct_thrd_create(&thr, &wait_thread, static_cast<void *>(argsptr.release())) != tct_thrd_success;
+
+}
+
+static SEXP execLater_fd_impl(const Rcpp::Function& callback, int num_fds, struct pollfd *fds, double timeout, int loop_id) {
+
+  std::shared_ptr<ThreadArgs> args = std::make_shared<ThreadArgs>(callback, num_fds, fds, timeout, loop_id);
+
+  if (execLater_launch_thread(args))
+    Rcpp::stop("Thread creation failed");
+
+  Rcpp::XPtr<std::shared_ptr<std::atomic<bool>>> xptr(new std::shared_ptr<std::atomic<bool>>(args->flag), true);
+  return xptr;
+
+}
+
+// native version
+static int execLater_fd_impl(void (*func)(int *, void *), void *data, int num_fds, struct pollfd *fds, double timeout, int loop_id) {
+
+  std::shared_ptr<ThreadArgs> args = std::make_shared<ThreadArgs>(func, data, num_fds, fds, timeout, loop_id);
+
+  return execLater_launch_thread(args);
+
+}
+
+// [[Rcpp::export]]
+Rcpp::RObject execLater_fd(Rcpp::Function callback, Rcpp::IntegerVector readfds, Rcpp::IntegerVector writefds,
+                           Rcpp::IntegerVector exceptfds, Rcpp::NumericVector timeoutSecs, Rcpp::IntegerVector loop_id) {
+
+  const int rfds = static_cast<int>(readfds.size());
+  const int wfds = static_cast<int>(writefds.size());
+  const int efds = static_cast<int>(exceptfds.size());
+  const int num_fds = rfds + wfds + efds;
+  const double timeout = num_fds ? timeoutSecs[0] : 0;
+  const int loop = loop_id[0];
+
+  std::vector<struct pollfd> pollfds;
+  pollfds.reserve(num_fds);
+  struct pollfd pfd;
+
+  for (int i = 0; i < rfds; i++) {
+    pfd.fd = readfds[i];
+    pfd.events = POLLIN;
+    pfd.revents = 0;
+    pollfds.push_back(pfd);
+  }
+  for (int i = 0; i < wfds; i++) {
+    pfd.fd = writefds[i];
+    pfd.events = POLLOUT;
+    pfd.revents = 0;
+    pollfds.push_back(pfd);
+  }
+  for (int i = 0; i < efds; i++) {
+    pfd.fd = exceptfds[i];
+    pfd.events = 0;
+    pfd.revents = 0;
+    pollfds.push_back(pfd);
+  }
+
+  return execLater_fd_impl(callback, num_fds, pollfds.data(), timeout, loop);
+
+}
+
+// [[Rcpp::export]]
+Rcpp::LogicalVector fd_cancel(Rcpp::RObject xptr) {
+
+  Rcpp::XPtr<std::shared_ptr<std::atomic<bool>>> flag(xptr);
+
+  if ((*flag)->load())
+    return false;
+
+  (*flag)->store(true);
+  return true;
```

Comment:
Another `compare_exchange`

Reply:
Oh, maybe doesn't matter (for correctness) in this case because you're always writing to it from the same thread.

Reply:
I agree with you. I've amened it anyway in [f9dbb70](https://github.com/r-lib/later/pull/190/commits/f9dbb7044b7a1ef549f3b53360a46060f746b097) as the intent is clearer. I also dislike the multiple return branches in this particular case.